src/codegen/statements.rs

```diff
@@ -106,6 +144,21 @@ pub(crate) fn statement(
         }
         Statement::Expression(_, reachable, expr) => {
             if let ast::Expression::Assign(_, _, left, right) = &expr {
+                // If the left operand is a variable of type Array
+                if let ast::Expression::Variable(_, Type::Array(_, _), pos) = &**left {
```

Comment:
This check should come after the `if should_remove_assignment`, otherwise you'll keep track of an assignment that does not exist in the code.